### ISSUE TITLE
[AST] bug for unpaired double quotes in regexps

### ISSUE THREAD

**Issue by jannick0** [jgm/pandoc#4316]
It appears that `pandoc` assumes that any encountered double quote within a regular expression (syntax like `/regular expression/`) signifies the beginning of a string and then waits for the next double quote to close the string unit. Thus, stand-alone double quotes, e.g. in regular expressions, provide unexpected (LaTeX) output. This applies to escaped as well as un-escaped double quotes in regular expressions.

**Example** (mini sample [here](https://github.com/jgm/pandoc/files/1670242/dq.zip)): 
```awk
BEGIN{
   # comment alone on the line
   a = "OK"
   # pandoc wrongly considers the double quote
   # in the regular expression (without pairing)
   # as beginning of string token and waits for
   # the next double quote to close the string token
   n = match(w, /[^"]/) # comment
   b = "I am a string"
}
```
LaTeX output with pandoc 2.0,6
```tex
\hypertarget{dq}{%
\label{dq}}%
\begin{Shaded}
\begin{Highlighting}[numbers=left,,]
\ControlFlowTok{BEGIN}\KeywordTok{\{}
   \CommentTok{# comment alone on the line}
\NormalTok{   a }\OperatorTok{=} \StringTok{"OK"}
   \CommentTok{# pandoc wrongly considers the double quote}
   \CommentTok{# in the regular expression (without pairing)}
   \CommentTok{# as beginning of string token and waits for}
   \CommentTok{# the next double quote to close the string token}
\NormalTok{   n }\OperatorTok{=} \FunctionTok{match}\NormalTok{(w}\OperatorTok{,} \OperatorTok{/}\NormalTok{[}\OperatorTok{^}\StringTok{"]/) # comment}
\StringTok{   b = "}\NormalTok{I am a string}\StringTok{"}
\StringTok{\}}
\end{Highlighting}
\end{Shaded}
```
Diff to expected output
```diff
--- dq_rx.tex	2018-01-27 17:23:57.730572100 +0100
+++ dq_rx_exp.tex	2018-01-27 17:23:51.376826300 +0100
@@ -9,8 +9,8 @@
    \CommentTok{# in the regular expression (without pairing)}
    \CommentTok{# as beginning of string token and waits for}
    \CommentTok{# the next double quote to close the string token}
-\NormalTok{   n }\OperatorTok{=} \FunctionTok{match}\NormalTok{(w}\OperatorTok{,} \OperatorTok{/}\NormalTok{[}\OperatorTok{^}\StringTok{"]/) # comment}
-\StringTok{   b = "}\NormalTok{I am a string}\StringTok{"}
+\NormalTok{   n }\OperatorTok{=} \FunctionTok{match}\NormalTok{(w}\OperatorTok{,} \OperatorTok{/}\NormalTok{[}\OperatorTok{^}\SpecialCharTok{"}\NormalTok{]}\OperatorTok{/}\NormalTok{) }\CommentTok{# comment}
+\NormalTok{   b = }\StringTok{"I am a string"}
 \StringTok{\}}
 \end{Highlighting}
 \end{Shaded}
```

**Comment by jgm** [comment]
This concerns syntax highlighting as `awk`.
Highlighting is done by jgm/skylighting, so that's the proper place to report a bug.
Note that the parser is derived from KDE/Kate's xml syntax definition for awk.  It's possible that the definition needs revising, or it may be a bug in skylighting.

You can test skylighing by itself by building it with the `executable` flag and running `skylighting -s awk -f native` on your awk sample.  I get this:
```
[ [ ( ControlFlowTok , "BEGIN" ) , ( KeywordTok , "{" ) ]
, [ ( NormalTok , "   " )
  , ( CommentTok , "# comment alone on the line" )
  ]
, [ ( NormalTok , "   a " )
  , ( OperatorTok , "=" )
  , ( NormalTok , " " )
  , ( StringTok , "\"OK\"" )
  ]
, [ ( NormalTok , "   " )
  , ( CommentTok , "# pandoc wrongly considers the double quote" )
  ]
, [ ( NormalTok , "   " )
  , ( CommentTok , "# in the regular expression (without pairing)" )
  ]
, [ ( NormalTok , "   " )
  , ( CommentTok , "# as beginning of string token and waits for" )
  ]
, [ ( NormalTok , "   " )
  , ( CommentTok
    , "# the next double quote to close the string token"
    )
  ]
, [ ( NormalTok , "   n " )
  , ( OperatorTok , "=" )
  , ( NormalTok , " " )
  , ( FunctionTok , "match" )
  , ( NormalTok , "(w" )
  , ( OperatorTok , "," )
  , ( NormalTok , " " )
  , ( OperatorTok , "/" )
  , ( NormalTok , "[" )
  , ( OperatorTok , "^" )
  , ( StringTok , "\"]/) # comment" )
  ]
, [ ( StringTok , "   b = \"" )
  , ( NormalTok , "I am a string" )
  , ( StringTok , "\"" )
  ]
, [ ( StringTok , "}" ) ]
]
```

**Comment by jgm** [comment]
Clearly
```
  , ( StringTok , "\"]/) # comment" )
```
seems wrong.  Please report this to jgm/skylighting and close it here.

**Comment by jannick0** [comment]
I would rather consider that a bug for KDE syntax highlighting and took it [there](https://bugs.kde.org/show_bug.cgi?id=389591). Downstream once resolved there hopefully soon. Just as a reference, I additionally [posted](https://bugs.kde.org/show_bug.cgi?id=389590) a syntax enhancement request for awk, too.

And **many** thanks for pointing me to `skylighting` which I could manage to compile (my first haskell compilation). I am understanding that `skylighting` does all the heavy work of code tokenizing and parsing and output creation (e.g. LaTeX output), such that I could have beed addressed issue #4314, too, but I was not aware before.

Please fell free to open an issue at jgm/skylighting with any references as you may find appropriate.

Thanks!
